pkg/asset/installconfig/aws/endpoints.go

```diff
@@ -175,6 +230,85 @@ func (s *Route53EndpointResolver) ResolveEndpoint(ctx context.Context, params ro
 	return route53.NewDefaultEndpointResolverV2().ResolveEndpoint(ctx, params)
 }
 
+// S3EndpointResolver implements EndpointResolverV2 interface for S3.
+type S3EndpointResolver struct {
+	*ServiceEndpointResolver
+}
+
+// ResolveEndpoint for S3.
+func (s *S3EndpointResolver) ResolveEndpoint(ctx context.Context, params s3.EndpointParameters) (smithyendpoints.Endpoint, error) {
+	params.UseDualStack = aws.Bool(s.endpointOptions.UseDualStack)
+	params.UseFIPS = aws.Bool(s.endpointOptions.UseFIPS)
+
+	// If custom endpoint not found, return default endpoint for the service.
+	endpoint, ok := s.endpoints[s3.ServiceID]
+	if !ok {
+		s.logger.Debug("Custom S3 endpoint not found, using default endpoint")
+		return s3.NewDefaultEndpointResolverV2().ResolveEndpoint(ctx, params)
+	}
+
+	s.logger.Debugf("Custom S3 endpoint found, using custom endpoint: %s", endpoint.URL)
+	params.Endpoint = aws.String(endpoint.URL)
+	params.Region = aws.String(s.endpointOptions.Region)
+	return s3.NewDefaultEndpointResolverV2().ResolveEndpoint(ctx, params)
+}
+
+// ELBV2EndpointResolver implements EndpointResolverV2 interface for ELBV2.
+type ELBV2EndpointResolver struct {
+	*ServiceEndpointResolver
+}
+
+// ResolveEndpoint for ELBV2.
+func (s *ELBV2EndpointResolver) ResolveEndpoint(ctx context.Context, params elbv2.EndpointParameters) (smithyendpoints.Endpoint, error) {
+	params.UseDualStack = aws.Bool(s.endpointOptions.UseDualStack)
+	params.UseFIPS = aws.Bool(s.endpointOptions.UseFIPS)
+
+	endpoint, ok := s.endpoints[elbv2.ServiceID]
+	if !ok {
+		// For backwards compatibility, we use the elb endpoint if
+		// no elbv2 endpoint is defined.
```

Comment:
Aren't elb and elbv2 technically two different services? There is an elb and elbv2 package in both v1 and v2, so I wonder if we need to alert that one wasn't defined not use the elb endpoint.

Reply:
Btw, upstream CAPA accepts this approach: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5680/files#diff-840355c57cb67a5b8dc01f54902be0aa7a76b760f7c522becc78fc37f42e2d0dR122-R134

That is if elbv2 endpoint is not specified, fall back to elb endpoint if any.